core/src/main/java/zipkin2/reporter/internal/AsyncReporter.java

```diff
@@ -216,20 +216,30 @@ void startFlusherThread() {
       flushThread.start();
     }
 
+    @SuppressWarnings("unchecked")
     @Override public void report(S next) {
       if (next == null) throw new NullPointerException("span == null");
       // Lazy start so that reporters never used don't spawn threads
       if (started.compareAndSet(false, true)) startFlusherThread();
       metrics.incrementSpans(1);
-      int nextSizeInBytes = encoder.sizeInBytes(next);
-      int messageSizeOfNextSpan = sender.messageSizeInBytes(nextSizeInBytes);
-      metrics.incrementSpanBytes(nextSizeInBytes);
-      if (closed.get() ||
-        // don't enqueue something larger than we can drain
-        messageSizeOfNextSpan > messageMaxBytes ||
-        !pending.offer(next, nextSizeInBytes)) {
-        metrics.incrementSpansDropped(1);
-      }
+      
+      if (pending instanceof UnsizedSpanConsumer) {
```

Comment:
this part feels uncomfortable, as it looks like we are moving details about the implementation into the call site

Reply:
added a commit to show genarlly what I was thinking about

Reply:
Thanks a lot for pushing the change, I was thinking towards reworking it in a similar way (but you already did it)

Reply:
another approach instead of the last commit is to make a super type of internal BoundedAsyncReporter which is about counts. in that type do the count related logic. make BoundedAsyncReporter extend that type and do the old logic which sizes the spans. return one or the other depending on count or size bounded. This may be better than my last commit as it probably handles blockhound concerns precisely.. also there's no if statements at runtime,

Reply:
I can look at it, sure

Reply:
@codefromthecrypt I think we should OK now: I removed the unnecessary methods (and consolidated a bit metrics collection in `drainTo`)